File: problems-adapt/2101-2200/2184_count-sturdy-brick-wall-layouts/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
  public:
    int sturdyWallLayouts(int height, int width, std::vector<int> &bricks) {
        // A row is fully described by its internal-joint bitmask; adjacent
        // rows must be disjoint. Enumerate row masks once, then run one
        // vector-of-counts transition per row.
        static const long long MOD = 1000000007LL;
        std::vector<int> masks;
        std::vector<int> current;
        struct Rec {
            static void go(int position, int width, const std::vector<int> &bricks, int mask, std::vector<int> &masks) {
                if (position == width) {
                    masks.push_back(mask);
                    return;
                }
                for (int brick : bricks) {
                    if (position + brick > width) {
                        continue;
                    }
                    int next = position + brick;
                    int extra = next < width ? (1 << (next - 1)) : 0;
                    go(next, width, bricks, mask | extra, masks);
                }
            }
        };
        Rec::go(0, width, bricks, 0, masks);
        if (masks.empty()) {
            return 0;
        }
        std::unordered_map<int, long long> counts;
        for (int mask : masks) {
            counts[mask] = 1;
        }
        for (int row = 1; row < height; ++row) {
            std::unordered_map<int, long long> nextCounts;
            for (int below : masks) {
                long long total = 0;
                for (int above : masks) {
                    if ((above & below) == 0) {
                        total += counts[above];
                    }
                }
                nextCounts[below] = total % MOD;
            }
            counts = std::move(nextCounts);
        }
        long long answer = 0;
        for (auto &entry : counts) {
            answer = (answer + entry.second) % MOD;
        }
        return static_cast<int>(answer);
    }
};

File: problems-adapt/2101-2200/2184_count-sturdy-brick-wall-layouts/solution.rs
impl Solution {
    pub fn sturdy_wall_layouts(height: i32, width: i32, bricks: Vec<i32>) -> i32 {
        // A row is fully described by its internal-joint bitmask; adjacent
        // rows must be disjoint. Enumerate row masks once, then run one
        // count-map transition per row.
        const MOD: i64 = 1_000_000_007;
        let mut masks: Vec<i32> = Vec::new();
        fn rec(position: i32, width: i32, bricks: &[i32], mask: i32, masks: &mut Vec<i32>) {
            if position == width {
                masks.push(mask);
                return;
            }
            for brick in bricks {
                if position + brick > width {
                    continue;
                }
                let next = position + brick;
                let extra = if next < width { 1 << (next - 1) } else { 0 };
                rec(next, width, bricks, mask | extra, masks);
            }
        }
        rec(0, width, &bricks, 0, &mut masks);
        if masks.is_empty() {
            return 0;
        }
        use std::collections::HashMap;
        let mut counts: HashMap<i32, i64> = masks.iter().map(|&m| (m, 1i64)).collect();
        for _row in 1..height {
            let mut next_counts: HashMap<i32, i64> = HashMap::new();
            for &below in &masks {
                let mut total = 0i64;
                for &above in &masks {
                    if above & below == 0 {
                        total += counts[&above];
                    }
                }
                next_counts.insert(below, total % MOD);
            }
            counts = next_counts;
        }
        let mut answer = 0i64;
        for value in counts.values() {
            answer = (answer + value) % MOD;
        }
        answer as i32
    }
}
